src/Compilers/CSharp/Portable/BoundTree/Expression.cs

```diff
@@ -0,0 +1,1791 @@
+﻿using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis.Semantics;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using System;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    partial class BoundExpression : IExpression
+    {
+        ITypeSymbol IExpression.ResultType
+        {
+            get { return this.Type; }
+        }
+
+        OperationKind IOperation.Kind
+        {
+            get { return this.ExpressionKind; }
+        }
+
+        object IExpression.ConstantValue
+        {
+            get
+            {
+                ConstantValue value = this.ConstantValue;
+                if (value == null)
+                {
+                    return null;
+                }
+
+                return value.Value;
+            }
+        }
+
+        SyntaxNode IOperation.Syntax
+        {
+            get { return this.Syntax; }
+        }
+
+        protected virtual OperationKind ExpressionKind { get { return OperationKind.None; } }
+        // protected abstract OperationKind ExpressionKind { get; }
+    }
+
+    internal abstract partial class BoundNode : IOperationSearchable
+    {
+        IEnumerable<IOperation> IOperationSearchable.Descendants()
+        {
+            var list = new List<BoundNode>();
+            new Collector(list).Visit(this);
+            list.RemoveAt(0);
+            return list.OfType<IOperation>();
+        }
+
+        IEnumerable<IOperation> IOperationSearchable.DescendantsAndSelf()
+        {
+            var list = new List<BoundNode>();
+            new Collector(list).Visit(this);
+            return list.OfType<IOperation>();
+        }
+
+        private class Collector : BoundTreeWalker
+        {
+            private readonly List<BoundNode> nodes;
+
+            public Collector(List<BoundNode> nodes)
+            {
+                this.nodes = nodes;
+            }
+
+            public override BoundNode Visit(BoundNode node)
+            {
+                this.nodes.Add(node);
+                return base.Visit(node);
+            }
+        }
+    }
+
+    partial class BoundCall : IInvocation
+    {
+        IMethodSymbol IInvocation.TargetMethod
+        {
+            get { return this.Method; }
+        }
+
+        IExpression IInvocation.Instance
+        {
+            get { return this.ReceiverOpt; }
+        }
+
+        InvocationKind IInvocation.InvocationKind
+        {
+            get
+            {
+                if (this.Method.IsStatic)
+                {
+                    return InvocationKind.Static;
+                }
+
+                if (this.Method.IsVirtual && !this.ReceiverOpt.SuppressVirtualCalls)
```

Comment:
I've found other code that explicitly checks for virtual/abstract/override, so I'll employ the three checks here.